File: Maximum Subarray/Maximum Subarray.java
public class Solution {
    public int maxSubArray(int[] A) {
        int max = A[0];
        int accu = (A[0] > 0) ? A[0] : 0;
        for(int i = 1; i < A.length; ++i){
            if(A[i] < 0){
                max = Math.max(max, A[i]);
                accu += A[i];
                accu = (accu > 0) ? accu : 0;
            }
            else{
                accu += A[i];
                max = Math.max(max, accu);
            }
        }
        
        return max;
    }
}

File: Maximum Subarray/Maximum Subarray.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def maxSubArray(self, A):
        if A is None or len(A) == 0:
            return 0
        maxVal = A[0]
        accu = A[0];
        if A[0] < 0:
            accu = 0
        
        for i in xrange(1, len(A)):
            if A[i] < 0:
                maxVal = max(maxVal, A[i])
                accu = accu + A[i]
                if accu < 0:
                    accu = 0
            else:
                accu = accu + A[i]
                maxVal = max(maxVal, accu)
                
        return maxVal
